pip/pip-297.md

```diff
@@ -0,0 +1,168 @@
+the notifyError method to the PushSource class in PIP-281: https://github.com/apache/pulsar/pull/20807. However, this
+does not solve the same problem that all source connectors face.
+
+The problem is same for the Pulsar Function. The user can also process messages asynchronously in a Pulsar Function. If
+any fatal exceptions happen in the user's thread, the function framework can't handle the exception
+correctly.
+
+We need a way for the connector and function developers to throw unrecoverable exceptions outside the function instance
+thread. The function framework should catch these exceptions and terminate the function accordingly.
+
+# Goals
+
+## In Scope
+
+- This proposal will apply both to the Pulsar Function and the Pulsar Connector.
+- Support throwing unrecoverable exceptions outside the function instance thread.
+- The function framework should terminate the connector or function when there are any unrecoverable exceptions thrown
+  asynchronously.
+
+## Out of Scope
+
+- The fixes of the exception-raising issue mentioned in the Motivation part for all the connectors are not included in
+  this PIP. This PIP only provides the feature for the connector developer to raise the exception outside the function
+  instance thread. The fixes should be in serval different PRs.
+
+# High Level Design
+
+Introduce a new method `raiseException` to the context. All the connector implementation code and the function code can
+use this context and call the `raiseException` method to raise an unrecoverable exception to the function framework.
+
+After the connector or function raises the unrecoverable exception, the function instance thread will be interrupted.
+The function framework then could catch the exception, log it, and then terminate the function instance.
+
+# Detailed Design
+
+## Design & Implementation Details
+
+This PIP proposes to add a new method`raiseException`to the context `BaseContext`. This method allows the
+connector or the function code to report an unrecoverable exception to the function framework. The `SinkContext`
+and `SourceContext`
+are all inherited from `BaseContext`. Therefore, all the sink connectors and source connectors can invoke this new
+method. The pulsar function context class `Context` is also inherited from `BaseContext`. Therefore, the function code
+can also invoke this new method.
+
+In the `raiseException`, the function instance thread will be interrupted. The function instance thread can then catch
+the interrupt exception and get the unrecoverable exception. The function framework then logs this exception,
+reports to the metrics, and finally terminates the function instance.
+
+## Public-facing Changes
+
+### Public API
+
+Introduce `raiseException` method to the `BaseContext`:
+
+```java
+public interface BaseContext {
+    /**
+     * Raise an unrecoverable exception to the function framework.
+     *
+     * @param e the exception to be raised
+     */
+    void raiseException(Exception e);
```

Comment:
let's use "Throwable" instead of Exception

Reply:
I think that it is better to have two methods:
- raiseUnrecoverableError
- raiseTemporaryError

This way it is clearer for the framework what it is expected to happen, the parameter is only for descriptive purpose, the framework cannot decided if a problem is temporary or definitive given the exception (for instance an IOException may mean both temporary and permanent failures)

Reply:
I think the PIP has introduced the `Recoverable Exception` and `UnRecoverable Exception` which confused the reader. I have refined the PIP to make it clearer for the exception handling.

For the `TemporaryError` here, we don't need to notify the function framework, the function and the connector should handle it by itself.